File: frontend/src/types/activityFeed.ts
/**
 * Activity Feed types for Feature 018 - Activity Stream Feed
 *
 * Represents activities from followed users (trip published, photo uploaded, achievement unlocked).
 */

/**
 * Activity type enum
 */
export type ActivityType = 'TRIP_PUBLISHED' | 'PHOTO_UPLOADED' | 'ACHIEVEMENT_UNLOCKED';

/**
 * Public user summary for activity author
 */
export interface PublicUserSummary {
  user_id: string;
  username: string;
  photo_url: string | null;
}

/**
 * Activity metadata (type-specific data)
 */
export interface ActivityMetadata {
  // TRIP_PUBLISHED metadata
  trip_id?: string; // Trip UUID for navigation
  trip_title?: string;
  trip_distance_km?: number;
  trip_photo_url?: string;

  // PHOTO_UPLOADED metadata
  photo_url?: string;

  // ACHIEVEMENT_UNLOCKED metadata
  achievement_name?: string;
  achievement_badge?: string;
}

/**
 * Activity feed item
 * Matches backend schema: ActivityFeedItemSchema (backend/src/schemas/feed.py)
 */
export interface ActivityFeedItem {
  activity_id: string;
  user: PublicUserSummary;
  activity_type: ActivityType;
  metadata: ActivityMetadata;
  created_at: string; // ISO 8601 datetime
  likes_count: number;
  comments_count: number;
  is_liked_by_me: boolean;
}

/**
 * Activity feed response with cursor-based pagination
 * Matches backend schema: ActivityFeedResponseSchema (backend/src/schemas/feed.py)
 */
export interface ActivityFeedResponse {
  activities: ActivityFeedItem[];
  next_cursor: string | null;
  has_next: boolean;
}

/**
 * Parameters for getActivityFeed()
 */
export interface GetActivityFeedParams {
  cursor?: string | null; // Cursor for pagination (null for first page)
  limit?: number; // Items per page (min 1, max 50, default 20)
}

File: backend/src/api/activity_feed.py
"""
Activity Feed API endpoints (Feature 018 - US1).

Provides chronological activity stream from followed users.
"""

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from src.api.deps import get_current_user, get_db
from src.models.user import User
from src.schemas.feed import ActivityFeedResponseSchema
from src.services.feed_service import FeedService

router = APIRouter(prefix="/activity-feed", tags=["Activity Feed"])


@router.get("", response_model=ActivityFeedResponseSchema)
async def get_activity_feed(
    limit: int = Query(default=20, ge=1, le=50, description="Items per page (min 1, max 50)"),
    cursor: str | None = Query(default=None, description="Cursor for pagination (null for first page)"),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> ActivityFeedResponseSchema:
    """
    Get activity feed from followed users (Feature 018 - US1, T027, T028).

    Returns chronological activity stream with cursor-based pagination.

    **Activity Types**:
    - TRIP_PUBLISHED: User published a new trip
    - PHOTO_UPLOADED: User uploaded photos to a trip
    - ACHIEVEMENT_UNLOCKED: User earned an achievement badge

    **Pagination**:
    - limit: min 1, max 50, default 20
    - cursor: Opaque cursor string from previous response (null for first page)

    **Returns**:
    - activities: Array of ActivityFeedItemSchema objects
    - next_cursor: Cursor for next page (null if last page)
    - has_next: True if more activities exist

    **Performance**: <2s for 20 activities (SC-001)

    **Authentication**: Required (JWT Bearer token)
    """
    result = await FeedService.get_user_feed(
        db=db,
        user_id=current_user.user_id,
        limit=limit,
        cursor=cursor,
    )

    return ActivityFeedResponseSchema(**result)
